### PR TITLE
Reduce code duplication in Points

### LINKED ISSUE #1136: Uncovered lines in points properties code
In this code report:

https://codecov.io/gh/napari/napari/pull/1069/diff

these lines are not covered:

https://github.com/napari/napari/blob/6ea95a9fa2f9150a4dbb5ec1286b8ff2020c3957/napari/layers/points/points.py#L565-L575

and

https://github.com/napari/napari/blob/6ea95a9fa2f9150a4dbb5ec1286b8ff2020c3957/napari/layers/points/points.py#L606-L608

(The "ifs" are executed but never evaluate to True in tests.) It's possible that the second one in particular will *never* be hit, since we automatically convert input to arrays. So the code could possibly be removed.

### PR DESCRIPTION
**Pull request by kevinyamauchi**
# Description
This PR aims to reduce some of the code duplication in the Points layer that was introduced when adding properties and allowing colors to be mapped to properties. The general approach is to write general functions that can be used for both edge_color and face_color properties using `setattr` and `getattr`.

I also added some tests to increase coverage on the Points model (closes #1136).

## Type of change
- [x] Bug-fix (non-breaking change which fixes an issue)

# References

# How has this been tested?
- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: This is over-indented.

Comment (jni) [reply napari/layers/points/points.py]: Twice over. :joy:

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        Parameters
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        Parameters
```

**Review by jni — COMMENTED** [review]
@kevinyamauchi I've only been able to quickly skim but on first pass this looks great! Let us know when it's out of WIP. =)

File: napari/layers/points/points.py
Comment (kevinyamauchi) [reply napari/layers/points/points.py]: 🤦 !

**Comment by kevinyamauchi** [comment]
Thanks for the initial comments, @jni! I think this is ready for full review!

Also, the tests need a similar refactor, but I will do that in a separate PR, so we aren't changing the tests and code at the same time.

**Review by sofroniewn — APPROVED** [review]
Looks good to me! Nice deduplication!!

File: napari/layers/points/_tests/test_points.py
Comment (jni) [inline napari/layers/points/_tests/test_points.py]:
```suggestion
    properties_list = {'point_type': ['A', 'B'] * (shape[0] // 2)}
```

Having said this, I feel this is a bit brittle, and will only work when `shape[0]` is even. I think the "correct" way here would be to use itertools:

```python
from itertools import cycle, islice

point_type = list(islice(cycle('AB'), None, shape[0]))
```

Or:

```python
from itertools import cycle
import toolz as tz

point_type = list(tz.take(shape[0], cycle('AB')))
```

Comment (jni) [reply napari/layers/points/_tests/test_points.py]: And you make this multiple times in this test so might as well make it once (as a list) and transform it to `np.array` where needed.

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        attribute : str in {'edge', 'face'}
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        """Initialize current_{edge,face}_color when starting with empty layer.
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        """Add the edge or face colors for new points.
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
        attribute : str in {'edge', 'face'}
```

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]:
```suggestion
                        f'_{attribute}_color_property',
```

Also, since you use this multiple times in this function, it's probably worth creating it once, e.g. `color_property_attr = f'_{attribute}_color_property'` then using that each time. But maybe not worth it, as the string is more explicit. Either way, use f-strings. =)

File: napari/layers/points/_tests/test_points.py
Comment (jni) [inline napari/layers/points/_tests/test_points.py]: Perhaps this should be `np.random.random`, since it's continuous values?

**Review by jni — APPROVED** [review]
@kevinyamauchi good stuff! I don't want to hold up merge, so I'm approving, but there are some code style changes I'd like to see before merging:

- the way the properties lists are defined in the tests is a bit brittle, I've suggested a more generic approach (bonus: works with any number of discrete choices, not just two)
- when an input can only have a small possible set of values, typically these are listed in the type annotation in the docstring.
- use f-strings rather than `%s`.

File: napari/layers/points/points.py
Comment (kevinyamauchi) [reply napari/layers/points/points.py]: My preference is to write the string each time, because as you said, it's more explicit. However, I will change to f-strings.

File: napari/layers/points/_tests/test_points.py
Comment (kevinyamauchi) [reply napari/layers/points/_tests/test_points.py]: Thanks for the suggestion. I've implemented it as a function using the islice method. That way we can make property arrays of different lengths, if needed.

File: napari/layers/points/_tests/test_points.py
Comment (kevinyamauchi) [reply napari/layers/points/_tests/test_points.py]: That's a good point. I've updated it so there is a continuous property (using `np.random.random`) and a categorical property.

**Comment by kevinyamauchi** [comment]
Thank you for the reviews, @sofroniewn and @jni ! I have addressed your comments. I will let this sit until tomorrow to see if there are any other comments and then merge!

Next up: reduce duplication in points tests.

**Pull request merged**
